File: packages/algo-go/solutions/algo-00.go
package solutions

func MostFrequent(arr []int) int {
	var ret int
	maxCount := -1
	counter := make(map[int]int, len(arr))
	for _, n := range arr {
		if _, ok := counter[n]; !ok {
			counter[n] = 1
		} else {
			counter[n]++
		}

		if counter[n] > maxCount {
			maxCount = counter[n]
			ret = n
		}
	}
	return ret
}

File: packages/algo-rust/src/algo_00.rs
use std::collections::HashMap;

fn most_frequent(arr: &[i32]) -> i32 {
    let mut ret = 0;
    let mut max_count = -1;
    let mut counter = HashMap::new();
    for n in arr {
        let counter = counter.entry(n).or_insert(0);
        *counter += 1;
        if *counter > max_count {
            max_count = *counter;
            ret = *n;
        }
    }
    ret
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_most_frequent() {
        assert_eq!(most_frequent(&[1, 2, 1, 2, 1]), 1);
    }
}
